python/torch_mlir/eager_mode/torch_mlir_tensor.py

```diff
@@ -2,103 +2,161 @@
+    and are expected to provide conversion mechanisms between their representation and torch.Tensor.
 
-    This class uses the _make_wrapper_subclass pattern to override __torch_dispatch__
-    in order to dispatch through torch-mlir instead of aten. Here we basically only unwrap and wrap
-    torch.Tensors. Most of the heavy lifting is done in the adjacent torch_mlir_dispatch module.
+    Here we only verify that inputs abide by current supported features of Torch-MLIR (contiguous memory and
+    strided tensor layout) and build the mlir module. Importantly, we also recover from any malfunctions in the
+    deriving classes and dispatch back to conventional PyTorch.
 
-    More documentation on how this pattern works can be found in this forum post
+    More documentation on how the __torch_dispatch__ pattern works can be found in this forum post
     https://dev-discuss.pytorch.org/t/what-and-why-is-torch-dispatch/557
     and this RFC
     https://github.com/pytorch/rfcs/blob/master/RFC-0001-torch-function-for-methods.md#process-followed-during-a-functionmethod-call
     and this repo with many examples
     https://github.com/albanD/subclass_zoo
     """
 
-    elem: torch.Tensor
+    elem: Any
 
     __slots__ = ["elem"]
 
-    @staticmethod
-    def __new__(cls, elem, *args, **kwargs):
-        r = torch.Tensor._make_wrapper_subclass(
-            cls,
-            elem.size(),
-            strides=elem.stride(),
-            storage_offset=elem.storage_offset(),
-            dtype=elem.dtype,
-            layout=elem.layout,
-            device=elem.device,
-            # Only float tensors can have gradients.
-            requires_grad=elem.dtype in {torch.float, torch.float32, torch.float64}
-            and (kwargs.get("requires_grad", False) or elem.requires_grad),
-        )
-        r.elem = elem.detach() if r.requires_grad else elem
+    def __new__(cls, elem, **kwargs):
+        """Wrap elem (which could be a torch.Tensor or otherwise) in a torch.Tensor subclass.
+
+        Critically, this method needs to parse relevant metadata from the device representation
+        (such as shape, striding, dtype, etc.) and translate it into torch conventions.
+
+        Deriving classes must provide a way to construct themselves from either their device specific representation
+        or torch.Tensor; the latter is to handle the case that dispatch to PyTorch to recover from an error.
+        """
+        if isinstance(elem, torch.Tensor):
+            r = make_wrapper_subclass_from_torch_tensor(cls, elem, **kwargs)
+            r.elem = backend.transfer_from_torch_to_device(elem)
+        elif isinstance(elem, torch.nn.Parameter):
+            r = make_wrapper_subclass_from_torch_tensor(cls, elem.data, **kwargs)
+            r.elem = backend.transfer_from_torch_to_device(elem.detach().data)
+        elif isinstance(elem, backend.device_array_type()):
```

Comment:
This doesn't feel right to me. We should always know when we are wrapping a DeviceTensor -- we shouldn't need to do a brittle check on the isinstance/etc.).

Reply:
The alternative is just to assume you've DeviceTensor if you've fallen through the previous two cases.

Reply:
@silvasean coming back to this: I'm not sure how we know this for a given `elem` since we can enter this constructor from, in principle, anywhere.